pkg/agent/cniserver/ipam/antrea_ipam.go

```diff
@@ -147,29 +147,47 @@ func (d *AntreaIPAM) Add(args *invoke.Args, k8sArgs *types.K8sArgs, networkConfi
 	}
 
 	owner := getAllocationOwner(args, k8sArgs, reservedOwner, false)
-	var ip net.IP
-	var subnetInfo *crdv1a2.SubnetInfo
-	if reservedOwner != nil {
-		ip, subnetInfo, err = allocator.AllocateReservedOrNext(crdv1a2.IPAddressPhaseAllocated, owner)
-	} else if len(ips) == 0 {
-		ip, subnetInfo, err = allocator.AllocateNext(crdv1a2.IPAddressPhaseAllocated, owner)
-	} else {
-		ip = ips[0]
-		subnetInfo, err = allocator.AllocateIP(ip, crdv1a2.IPAddressPhaseAllocated, owner)
-	}
-	if err != nil {
-		return true, nil, err
-	}
+	result := IPAMResult{Result: current.Result{CNIVersion: current.ImplementedSpecVersion}}
+	isSuccessful := false
+	for i := range allocators {
+		var ip net.IP
+		var subnetInfo *crdv1a2.SubnetInfo
+		allocator := allocators[i]
+		if reservedOwner != nil {
+			ip, subnetInfo, err = allocator.AllocateReservedOrNext(crdv1a2.IPAddressPhaseAllocated, owner)
+		} else if len(ips) == 0 {
+			ip, subnetInfo, err = allocator.AllocateNext(crdv1a2.IPAddressPhaseAllocated, owner)
+		} else {
+			ip = ips[i]
+			subnetInfo, err = allocator.AllocateIP(ip, crdv1a2.IPAddressPhaseAllocated, owner)
+		}
+		if err != nil {
+			return true, nil, err
+		}
+		if i == 0 {
+			result.VLANID = subnetInfo.VLAN
+		} else if result.VLANID != subnetInfo.VLAN {
+			return true, nil, fmt.Errorf("VLAN in IP pools is not same")
+		}
+		rollbackFunc := func() {
+			if !isSuccessful {
+				if err := allocator.Release(ip); err != nil {
```

Comment:
When the func is called, the value of "allocator" will be the last failed allocator, but not the earlier succeeded ones?

Reply:
Nice question. We use `allocator` instead of `allocators[i]` to solve this problem, I'll add comments and UT for this.

Reply:
Use ReleaseContainer instead of Release IP to avoid unexpected deletion.
Hit some issue on UT if we release IP before informer synced IPPool updates, and added retry to ReleaseContainer func.